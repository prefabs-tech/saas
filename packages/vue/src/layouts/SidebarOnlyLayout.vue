<template>
  <Layout :menu="menu" :no-locale-switcher="noLocaleSwitcher">
    <template #userMenu>
      <UserMenu v-if="showUserMenu" />
    </template>

    <template #afterNavLinks>
      <slot name="afterNavLinks"></slot>
    </template>

    <template #afterSidebarMenu>
      <slot name="afterSidebarMenu"></slot>
    </template>

    <template #locales>
      <slot name="locales"></slot>
    </template>

    <slot></slot>
  </Layout>
</template>

<script lang="ts">
export default {
  name: "SidebarOnlyLayout",
};
</script>

<script setup lang="ts">
import { useConfig } from "@dzangolab/vue3-config";
import { SidebarOnlyLayout as Layout } from "@dzangolab/vue3-layout";
import { storeToRefs } from "pinia";
import { computed } from "vue";
import { useRouter } from "vue-router";

import UserMenu from "../components/UserMenu.vue";
import useUserStore from "../store";

import type { MenuItem, SidebarMenu } from "@dzangolab/vue3-layout";

defineProps({
  noLocaleSwitcher: Boolean,
  showUserMenu: {
    default: true,
    required: false,
    type: Boolean,
  },
});

const userStore = useUserStore();

const { user } = storeToRefs(userStore);

const { layout: layoutConfig } = useConfig();

const router = useRouter();

const allRoutes = router.getRoutes();

const menu = computed(() => {
  let menuItems = layoutConfig?.mainMenu || [];

  if (!user.value) {
    menuItems = menuItems?.filter((item: MenuItem) => {
      const route = allRoutes.find((r) => {
        if (r.name === item.route) {
          return true;
        }

        if (item.children?.length) {
          return (item.children as MenuItem[]).some((child: MenuItem) => {
            return allRoutes.some((cr) => cr.name === child.route);
          });
        }

        return false;
      });

      return route && !route.meta?.authenticated;
    }) as MenuItem[];
  }

  return menuItems.map((item: MenuItem) => {
    return {
      hide: item?.hide,
      icon: item?.icon,
      name: item.name,
      routeName: item.route,
      shortName: item?.shortName,
      children: item?.children?.map((childItem: MenuItem) => {
        return {
          hide: childItem?.hide,
          icon: childItem?.icon,
          name: childItem.name,
          routeName: childItem.route,
          shortName: childItem?.shortName,
        };
      }),
    };
  }) as SidebarMenu[];
});
</script>

<style lang="css">
.sidebar-only > .sidebar .user-menu-dropdown.expanded .toggle > svg {
  transform: rotate(-180deg);
}

.sidebar-only > .sidebar .user-menu-dropdown .toggle > svg {
  transform: rotate(0);
}

.sidebar-only .user-menu > ul {
  flex-direction: column;
}

.sidebar-only .user-menu > ul > li,
.sidebar-only .user-menu-dropdown {
  --_font-size: var(--font-size-min, 0.8rem);
  --_font-weight: var(--font-weight, 450);
  --dropdown-container-bg-color: #0870e5;
  --menu-highlight-color: #0870e5;

  font-size: var(--_font-size);
  font-weight: var(--_font-weight);
  width: 100%;
}

.sidebar-only .user-menu > ul > li:has(.router-link-exact-active) {
  background-color: #0870e5;
}

.sidebar-only .user-menu:not(.user-menu-dropdown) > ul > li > a,
.sidebar-only .user-menu-dropdown,
.sidebar-only .user-menu-dropdown > ul.dropdown > li {
  --_height: var(--nav-menu-height, 3rem);

  align-items: center;
  display: flex;
  height: var(--_height);
  width: 100%;
}

.sidebar-only .user-menu-dropdown:hover {
  background-color: #0870e5;
}

.sidebar-only .user-menu-dropdown.expanded > ul > li:hover {
  --dropdown-bg-color-hover: #0870e5;
}

.sidebar-only .user-menu-dropdown > .dropdown {
  opacity: 0;
  transform: translate3d(0, 0, 0);
  transition:
    transform 0.3s ease,
    opacity 0.5s ease;
}

.sidebar-only .user-menu-dropdown.expanded > .dropdown {
  --dropdown-bg-color: #007aff;
  --dropdown-border: 1px solid #007aff;

  box-shadow: 0 -2px 10px 2px #0870e5;
  opacity: 1;
  transform: translate3d(0, -152.5px, 0);
  width: 100%;
}
</style>
