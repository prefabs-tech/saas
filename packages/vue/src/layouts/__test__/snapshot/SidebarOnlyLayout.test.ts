import configPlugin from "@dzangolab/vue3-config";
import { mount } from "@vue/test-utils";
import { createPinia } from "pinia";
import { describe, expect, it } from "vitest";
import { createRouter, createWebHistory } from "vue-router";

import SidebarOnlyLayout from "../../SidebarOnlyLayout.vue";
import appConfig from "../config";

import type { VueWrapper } from "@vue/test-utils";

describe("SidebarOnlyLayout", () => {
  it("matches snapshot", () => {
    const pinia = createPinia();
    const router = createRouter({
      history: createWebHistory(),
      routes: [
        {
          path: "/",
          name: "home",
          component: SidebarOnlyLayout,
        },
      ],
    });

    const wrapper: VueWrapper = mount(SidebarOnlyLayout, {
      global: {
        plugins: [
          pinia,
          [
            configPlugin,
            {
              config: appConfig,
            },
          ],
          router,
        ],
        stubs: {
          UserMenu: true,
          LocaleSwitcher: true,
        },
      },
      props: {
        menu: [
          {
            name: "Menu1",
            routeName: "menu-1",
            shortName: "M1",
          },
          {
            name: "Menu2",
            routeName: "menu-2",
            shortName: "M2",
          },
        ],
      },
      slots: {
        afterNavLinks: "<div class='after-nav'></div>",
        afterSidebarMenu: "<div class='after-sidebar'></div>",
        default: "<div class='default'></div>",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
