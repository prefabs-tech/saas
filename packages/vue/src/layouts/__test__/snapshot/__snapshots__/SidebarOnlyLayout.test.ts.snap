// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SidebarOnlyLayout > matches snapshot 1`] = `
<div
  class="layout sidebar-only"
>
  <div
    class="sidebar layout-sidebar"
  >
    <div
      class="header"
    >
      <div
        class="logo"
      >
        <a
          class="logo"
          href="/"
        >
          <img
            alt="dzango"
            src="/logo.png"
          />
        </a>
      </div>
      <div
        class="title"
      >
        
        
      </div>
      <div
        class="toggle"
      >
        <transition-stub
          appear="false"
          css="true"
          name="sidebar-toggle"
          persisted="false"
        >
          <img
            alt="extend sidebar"
            class="extend"
            src="data:image/svg+xml,%3csvg%20viewBox='0%200%2024%2024'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cg%20stroke-width='0'%3e%3c/g%3e%3cg%20stroke-linecap='round'%20stroke-linejoin='round'%3e%3c/g%3e%3cg%3e%3cpath%20fill-rule='evenodd'%20clip-rule='evenodd'%20d='M8.29289%204.29289C8.68342%203.90237%209.31658%203.90237%209.70711%204.29289L16.7071%2011.2929C17.0976%2011.6834%2017.0976%2012.3166%2016.7071%2012.7071L9.70711%2019.7071C9.31658%2020.0976%208.68342%2020.0976%208.29289%2019.7071C7.90237%2019.3166%207.90237%2018.6834%208.29289%2018.2929L14.5858%2012L8.29289%205.70711C7.90237%205.31658%207.90237%204.68342%208.29289%204.29289Z'%20fill='%23ffffff'%3e%3c/path%3e%3c/g%3e%3c/svg%3e"
          />
        </transition-stub>
      </div>
    </div>
    <div
      class="sidebar-menu-wrapper"
    >
      
      
      <ul
        class="nav"
      >
        
        <nav
          class="nav-menu"
        >
          <div
            class="menu"
          >
            
            <div
              class="nav-menu-item"
            >
              <a
                aria-label="open menu"
                class="link"
              >
                <!---->
                <span>
                  M1
                </span>
                <!---->
                <!---->
              </a>
              <transition-stub
                appear="false"
                css="true"
                name="fade"
                persisted="false"
              >
                <!---->
              </transition-stub>
            </div>
            <div
              class="nav-menu-item"
            >
              <a
                aria-label="open menu"
                class="link"
              >
                <!---->
                <span>
                  M2
                </span>
                <!---->
                <!---->
              </a>
              <transition-stub
                appear="false"
                css="true"
                name="fade"
                persisted="false"
              >
                <!---->
              </transition-stub>
            </div>
            
          </div>
        </nav>
        
        
        
        
        <div
          class="after-nav"
        />
        
        
        
      </ul>
      
      
      
      <div
        class="after-sidebar"
      />
      
      
      
      <user-menu-stub />
      
      
      <locale-switcher-stub
        class="locales"
      />
      
      
    </div>
    <div
      class="footer"
    >
      
      <footer>
        
        <!---->
        
        
        
        
        <div
          class="version"
        >
          0.1.0
        </div>
        
      </footer>
      
    </div>
  </div>
  <main>
    
    
    <div
      class="default"
    />
    
    
  </main>
</div>
`;
